ngraph_bridge/ngraph_executor_db.cc

```diff
@@ -0,0 +1,135 @@
+/*******************************************************************************
+ * Copyright 2019 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+
+#include "ngraph_bridge/ngraph_executor_db.h"
+
+namespace tensorflow {
+
+namespace ngraph_bridge {
+
+NGraphExecutorDB::~NGraphExecutorDB() {
+  m_executable_pipelined_tensors_map.clear();
+  m_ng_function_map.clear();
+  m_ng_exec_map.clear();
+}
+
+bool NGraphExecutorDB::MaybeGetNgExecutable(
+    std::string signature,
+    std::shared_ptr<ngraph::runtime::Executable>& ng_exec) {
+  lock_guard<mutex> lock(m_mutex);
+  auto it = m_ng_exec_map.find(signature);
+  if (it == m_ng_exec_map.end()) {
+    return false;
+  }
+  ng_exec = it->second;
+  UpdateLRU(signature);
+  return true;
+}
+
+void NGraphExecutorDB::AddItem(
+    std::string signature,
+    std::pair<std::shared_ptr<ngraph::runtime::Executable>,
+              std::shared_ptr<ngraph::Function>>
+        ng_exec_func,
+    std::shared_ptr<ngraph::runtime::Executable>& evicted_ng_exec, int depth) {
+  auto ng_exec = ng_exec_func.first;
+  auto ng_function = ng_exec_func.second;
+  if (!signature.empty() && ng_function && ng_exec) {
+    lock_guard<mutex> lock(m_mutex);
+    const char* cache_depth_specified =
+        std::getenv("NGRAPH_TF_FUNCTION_CACHE_ITEM_DEPTH");
+    if (cache_depth_specified != nullptr) {
+      if (m_ng_exec_map.size() >= atoi(cache_depth_specified)) {
+        RemoveItem(m_lru.back(), evicted_ng_exec);
+      }
+    }
+    auto ng_exec_map_output_pair = m_ng_exec_map.emplace(signature, ng_exec);
+    if (ng_exec_map_output_pair.second) {
```

Comment:
Like we discussed, this is similar to `if(! MaybeGetNgExecutable)`. We could discuss more about this tomorrow